src/webgpu/util/texture/layout.ts

```diff
@@ -0,0 +1,153 @@
+import { C, assert } from '../../../common/framework/index.js';
+import { kTextureFormatInfo } from '../../capability_info.js';
+
+export const kBytesPerRowAlignment = 256;
+
+export function getTextureCopyLayout(
+  format: GPUTextureFormat,
+  size: GPUExtent3D,
+  mipLevel: number = 0,
+  bytesPerRow?: number,
+  rowsPerImage?: number
+): {
+  bytesPerTexel: number;
+  byteLength: number;
+  minBytesPerRow: number;
+  bytesPerRow: number;
+  rowsPerImage: number;
+} {
+  if (!Array.isArray(size)) {
+    size = [size.width, size.height, size.depth];
+  }
+
+  const bytesPerTexel = kTextureFormatInfo[format].bytes;
+  const minBytesPerRow = (size[0] >> mipLevel) * bytesPerTexel;
+  const alignedMinBytesPerRow =
+    Math.ceil(minBytesPerRow / kBytesPerRowAlignment) * kBytesPerRowAlignment;
+  if (bytesPerRow !== undefined) {
+    assert(bytesPerRow >= alignedMinBytesPerRow);
+  } else {
+    bytesPerRow = alignedMinBytesPerRow;
+  }
+
+  if (rowsPerImage !== undefined) {
+    assert(rowsPerImage >= size[1]);
+  } else {
+    rowsPerImage = size[1];
+  }
+
+  return {
+    bytesPerTexel,
+    byteLength: bytesPerRow * rowsPerImage,
+    minBytesPerRow,
+    bytesPerRow,
+    rowsPerImage,
+  };
+}
+
+export function fillTextureDataRows(
```

Comment:
nit: didn't realize at first this was filling every texel with the same data (instead of every row), maybe `fillTextureDataWithTexelValue`

Reply:
Done